segmentstore/server/src/main/java/io/pravega/segmentstore/server/reading/StreamSegmentReadIndex.java

```diff
@@ -200,9 +200,12 @@ public long updateGenerations(int currentGeneration, int oldestGeneration) {
                 // We can only evict if both these conditions are met:
                 // 1. The entry is a Cache Entry (Redirect entries cannot be removed).
                 // 2. Every single byte in the entry has to exist in Storage.
+                // In addition, we are free to evict (regardless of Generation, but still subject to the above rules) if
+                // every single byte in the entry has been truncated out.
+                long lastOffset = entry.getLastStreamSegmentOffset();
                 boolean canRemove = entry.isDataEntry()
-                        && entry.getLastStreamSegmentOffset() <= this.metadata.getStorageLength()
-                        && entry.getGeneration() < oldestGeneration;
+                        && lastOffset <= this.metadata.getStorageLength()
```

Comment:
Should it be instead `lastOffset < this.metadata.getStorageLength()`?

Reply:
lastOffset == StartOffset + Length, which means `<=` is appropriate. It means the length of the Segment up to the end of that Entry.